File: Patterns/Command Pattern/Cpp/RemoteControl.cpp
#pragma once

#include <iostream>
#include <vector>
#include <typeinfo>

#include "./Command/ICommand.cpp"
#include "./Command/NoCommand.cpp"

using namespace std;

class RemoteControl {
private:
    vector<ICommand*> onCommands;
    vector<ICommand*> offCommands;
    ICommand* undoCommand;

public:
    RemoteControl() {
        onCommands = vector<ICommand*>(7);
        offCommands = vector<ICommand*>(7);

        ICommand* noCommand = new NoCommand();
        for (int i = 0; i < 7; i++) {
            onCommands[i] = offCommands[i] = noCommand;
        }
        undoCommand = noCommand;
    }

    void setCommand(int slot, ICommand* onCommand, ICommand* offCommand) {
        onCommands[slot] = onCommand;
        offCommands[slot] = offCommand;
    }

    void onButtonWasPushed(int slot) {
        onCommands[slot]->execute();
        undoCommand = onCommands[slot];
    }

    void offButtonWasPushed(int slot) {
        offCommands[slot]->execute();
        undoCommand = offCommands[slot];
    }

    void undoWasPushed() {
        undoCommand->undo();
    }

    void print() {
        cout << "\n------ 리모컨 ------\n";
        for (int i = 0; i < onCommands.size(); i++) {
            cout << "[slot " << i << "] " << onCommands[i]->getCommandName() << ' ' << offCommands[i]->getCommandName() << '\n';
        }
        cout << "[undo] " << undoCommand->getCommandName() << '\n';
        cout << '\n';
    }
};

File: Patterns/Command Pattern/CSharp/RemoteControl.cs
using System;

using ICommand = Command.ICommand;
using NoCommand = Command.NoCommand;

public class RemoteControl
{
    private ICommand[] onCommands;
    private ICommand[] offCommands;
    private ICommand undoCommand;

    public RemoteControl()
    {
        onCommands = new ICommand[7];
        offCommands = new ICommand[7];

        ICommand noCommand = new NoCommand();
        for (var i = 0; i < 7; i++)
        {
            onCommands[i] = offCommands[i] = noCommand;
        }
        undoCommand = noCommand;
    }

    public void SetCommand(int slot, ICommand onCommand, ICommand offCommand)
    {
        onCommands[slot] = onCommand;
        offCommands[slot] = offCommand;
    }

    public void OnButtonWasPushed(int slot)
    {
        onCommands[slot].Execute();
        undoCommand = onCommands[slot];
    }

    public void OffButtonWasPushed(int slot)
    {
        offCommands[slot].Execute();
        undoCommand = offCommands[slot];
    }

    public void UndoWasPushed() => undoCommand.Undo();

    public void Print()
    {
        Console.WriteLine("\n------ 리모컨 ------");
        for (var i = 0; i < 7; i++)
        {
            Console.WriteLine($"[slot {i}] {onCommands[i].GetType().Name} {offCommands[i].GetType().Name}");
        }
        Console.WriteLine($"[undo] {undoCommand.GetType().Name}\n");
    }
}